File: HRMS/HRMS.Web/Modules/ServerTypes/Operations/PayrollForm.ts
import { DecimalEditor, EnumEditor, initFormType, IntegerEditor, LookupEditor, PrefixedContext } from "@serenity-is/corelib";
import { Month } from "./Month";

export interface PayrollForm {
    EmployeeId: LookupEditor;
    Month: EnumEditor;
    Year: IntegerEditor;
    BasicSalary: DecimalEditor;
    Allowances: DecimalEditor;
    Deductions: DecimalEditor;
    NetSalary: DecimalEditor;
}

export class PayrollForm extends PrefixedContext {
    static readonly formKey = 'Operations.Payroll';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!PayrollForm.init) {
            PayrollForm.init = true;

            var w0 = LookupEditor;
            var w1 = EnumEditor;
            var w2 = IntegerEditor;
            var w3 = DecimalEditor;

            initFormType(PayrollForm, [
                'EmployeeId', w0,
                'Month', w1,
                'Year', w2,
                'BasicSalary', w3,
                'Allowances', w3,
                'Deductions', w3,
                'NetSalary', w3
            ]);
        }
    }
}

[Month]; // referenced types

File: HRMS/HRMS.Web/Modules/Operations/Payroll/PayrollForm.cs
using Serenity.ComponentModel;
using HRMS.Operations;

namespace HRMS.Operations.Forms;

[FormScript("Operations.Payroll")]
[BasedOnRow(typeof(PayrollRow), CheckNames = true)]
public class PayrollForm
{
    public int EmployeeId { get; set; }
    [HalfWidth]
    public Month Month { get; set; }
    [HalfWidth]
    public int Year { get; set; }
    public decimal BasicSalary { get; set; }
    public decimal Allowances { get; set; }
    public decimal Deductions { get; set; }
    public decimal NetSalary { get; set; }
}
